desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
+    -------
+    output : tuple
+        (pitch, intersect, is_bp, bp1, bp2).
+
+    """
+    if pitch is None:
+        if err and not original:
+            raise ValueError("No pitch values were given.")
+        return original
+    else:
+        pitch = jnp.atleast_1d(pitch)
+        intersect, is_bp, bp1, bp2 = _get_bounce_points(pitch, zeta, poly_B, poly_B_z)
+        return pitch, intersect, is_bp, bp1, bp2
+
+
+def bounce_integral(
+    eq,
+    rho=None,
+    alpha=None,
+    zeta_max=10 * jnp.pi,
+    pitch=None,
+    resolution=20,
+    method="tanh_sinh",
+):
+    """Returns a method to compute the bounce integral of any quantity.
+
+    The bounce integral is defined as F_ℓ(λ) = ∫ f(ℓ) / √(1 − λ |B|) dℓ, where
+        dℓ parameterizes the distance along the field line,
+        λ is a constant proportional to the magnetic moment over energy,
+        |B| is the norm of the magnetic field,
+        f(ℓ) is the quantity to integrate along the field line,
+        and the endpoints of the integration are at the bounce points.
+    For a particle with fixed λ, bounce points are defined to be the location
+    on the field line such that the particle's velocity parallel to the
+    magnetic field is zero, i.e. λ |B| = 1.
+
+    The bounce integral is defined up to a sign.
+    We choose the sign that corresponds the particle's guiding center trajectory
+    traveling in the direction of increasing field-line-following label.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium on which the bounce integral is defined.
+    rho : ndarray
+        Unique flux surface label coordinates.
+    alpha : ndarray
+        Unique field line label coordinates over a constant rho surface.
+    zeta_max : float
+        Max value for field line following coordinate.
+    pitch : ndarray
+        λ values to evaluate the bounce integral at.
+    resolution : int
+        Number of interpolation points (knots) used for splines in the quadrature.
+        A maximum of three bounce points can be detected in between knots.
+        The accuracy of the quadrature will increase as some function of
+        the number of knots over the number of detected bounce points.
+        So for well-behaved magnetic fields increasing resolution should increase
+        the accuracy of the quadrature.
+    method : str
+        The quadrature scheme used to evaluate the integral.
+        The "direct" method exactly integrates a cubic spline of the integrand.
+        The "tanh_sinh" method performs a Tanh-sinh quadrature, where independent cubic
+        splines are used for components in the integrand so that the singularity near
+        the bounce points can be captured more accurately than can be represented by a
+        polynomial.
+
+    Returns
+    -------
+    bi : callable
+        This callable method computes the bounce integral F_ℓ(λ) for every
+        specified field line ℓ (constant rho and alpha), for every λ value in
```

Comment:
what is the use case for specifying a different pitch to `bi` than to the `bounce_integral` function?

Reply:
I didn't know what the better api was. If someone wants to compute the bounce integral of many quantities with the same pitch values then specifying them early is better. If the pitch values change, then one can still use the same bounce integral function without redoing the coordinate mapping.